pkg/exemplars/proxy.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package exemplars
+
+import (
+	"context"
+	"io"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/exemplars/exemplarspb"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// Proxy implements exemplarspb.Exemplars gRPC that fanouts requests to
+// given exemplarspb.Exemplars and de-duplication on the way.
+type Proxy struct {
+	logger    log.Logger
+	exemplars func() []exemplarspb.ExemplarsClient
+}
+
+// RegisterExemplarsServer register exemplars server.
+func RegisterExemplarsServer(exemplarsSrv exemplarspb.ExemplarsServer) func(*grpc.Server) {
+	return func(s *grpc.Server) {
+		exemplarspb.RegisterExemplarsServer(s, exemplarsSrv)
+	}
+}
+
+// NewProxy return new exemplars.Proxy.
+func NewProxy(logger log.Logger, exemplars func() []exemplarspb.ExemplarsClient) *Proxy {
+	return &Proxy{
+		logger:    logger,
+		exemplars: exemplars,
+	}
+}
+
+type exemplarsStream struct {
+	client  exemplarspb.ExemplarsClient
+	request *exemplarspb.ExemplarsRequest
+	channel chan<- *exemplarspb.ExemplarData
+	server  exemplarspb.Exemplars_ExemplarsServer
+}
+
+func (s *Proxy) Exemplars(req *exemplarspb.ExemplarsRequest, srv exemplarspb.Exemplars_ExemplarsServer) error {
+	var (
+		g, gctx   = errgroup.WithContext(srv.Context())
+		respChan  = make(chan *exemplarspb.ExemplarData, 10)
+		exemplars []*exemplarspb.ExemplarData
+	)
+
+	for _, exemplarsClient := range s.exemplars() {
+		es := &exemplarsStream{
+			client:  exemplarsClient,
+			request: req,
+			channel: respChan,
+			server:  srv,
+		}
+		g.Go(func() error { return es.receive(gctx) })
+	}
+
+	go func() {
+		_ = g.Wait()
+		close(respChan)
+	}()
+
+	for resp := range respChan {
+		exemplars = append(exemplars, resp)
+	}
+
+	if err := g.Wait(); err != nil {
+		level.Error(s.logger).Log("err", err)
+		return err
+	}
+
+	for _, e := range exemplars {
```

Comment:
It looks like the proxy is not deduplicating at all. Should we deduplicate once we've collected all of the responses?